### PR TITLE
Downstream socket full duplex bugfix

### PR DESCRIPTION
**Pull request by aviramc**
The suggested patch for issue #478

### CODE REVIEW

**Comment by agentzh** [comment]
@aviramc Will you mind adding one or a few test cases to the current test suite to exercise this bug?

**Comment by aviramc** [comment]
@agentzh , I've added the required test case (verified that it fails without the fix)

File: src/ngx_http_lua_common.h [hunk 1]
```diff
@@ -382,8 +382,10 @@ typedef struct ngx_http_lua_ctx_s {
 
     ngx_int_t                exit_code;
 
-    ngx_http_lua_co_ctx_t   *downstream_co_ctx; /* co ctx for the coroutine
-                                                   reading the request body */
+    ngx_http_lua_co_ctx_t   *downstream_read_co_ctx; /* co ctx for the coroutine
+                                                        reading the request body */
+    ngx_http_lua_co_ctx_t   *downstream_write_co_ctx; /* co ctx for the coroutine
```
Comment (agentzh) [inline src/ngx_http_lua_common.h]: Hmm, seems like we can save some space by reusing the existing `read_co_ctx` and `write_co_ctx` fields here? We already have the `body_downstream` and `raw_downstream` flags to identify downstream sockets anyway :)

File: t/131-duplex-req-socket.t [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+               local data, err, partial
+               -- First we receive in a blocking fashion so that ctx->downstream_co_ctx will be changed
+               data, err, partial = req_socket:receive(1)
+               if err ~= "timeout" then
+                  ngx.log(ngx.ERR, "Did not get timeout in the receiving thread!")
+                  return
+               end
+            
+               -- Now, sleep so that coctx->data is changed to sleep handler
+               ngx.sleep(3)
```
Comment (agentzh) [inline t/131-duplex-req-socket.t]: Hmm, 3 seconds is quite some time. Can we reduce it a bit? Like 1 or 2 seconds?

File: t/131-duplex-req-socket.t [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+repeat_each(1);
+
+plan tests => repeat_each() * (blocks() * 2);
+
+no_shuffle();
+run_tests();
+
+__DATA__
+
+=== TEST 18: downstream cosocket used in two different threads. See issue #481
```
Comment (agentzh) [inline t/131-duplex-req-socket.t]: Test number is wrong. "18" looks like a copy&paste mistake :)

File: t/131-duplex-req-socket.t [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            
+            writer_thread = ngx.thread.spawn(writer, req_socket)
+            reader_thread = ngx.thread.spawn(reader, req_socket)
+            
+            ngx.thread.wait(writer_thread)
+            ngx.thread.wait(reader_thread)
+            ngx.log(ngx.INFO, "The two threads finished")
+';
+        }
+--- raw_request eval
```
Comment (agentzh) [inline t/131-duplex-req-socket.t]: Maybe we can just use the following instead?

```
--- request
POST /test
--- more_headers
Content-Length: 1
```

It looks simpler to me :)

**Comment by agentzh** [comment]
@aviramc Please look at my comments above. Thank you very much for your hard work. And I'm sorry for the long delay on my side.

File: t/131-duplex-req-socket.t [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+               local bytes, err
+               -- send in a slow manner with a low timeout, so that the timeout handler will be
+               bytes, err = req_socket:send("slow!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!")
+               if err ~= "timeout" then
+                  ngx.log(ngx.ERR, "Did not get timeout in the sending thread!")
+                  return
+               end
+            end
+            
+            local req_socket, err = ngx.req.socket(true)
```
Comment (agentzh) [inline t/131-duplex-req-socket.t]: This test case the "raw" downstream cosocket. Can we add another test case for the more common "body" downstream cosocket? Thanks!

**Comment by agentzh** [comment]
@aviramc Never mind, I've just committed a massively modified version of your patch to git master as commit 9e0503b03. Will you review and try it out on your side? I'd like to make a new release of ngx_lua very soon. Many thanks!

**Comment by aviramc** [comment]
@agentzh , sorry for the late reply. Looks like an elegant solution for the problem. Soon I'll incorporate this into my code, but this looks like this would work with no problems.

**Comment by agentzh** [comment]
@aviramc Cool, thanks!

**Pull request closed without merging**